torch/distributed/_mesh_layout.py

```diff
@@ -0,0 +1,75 @@
+"""
+Definition of CuTe inspired Layouts for DeviceMesh internal bookkeeping and functions to manipulate them
+"""
+
+import math
+from collections.abc import Iterator
+from dataclasses import dataclass
+from typing import TypeAlias
+
+from torch.distributed._pycute import (
+    coalesce,
+    complement,
+    composition,
+    flatten,
+    is_tuple,
+    Layout,
+)
+
+
+NestedIntTuple: TypeAlias = tuple["int | NestedIntTuple", ...]
+
+
+@dataclass(frozen=True, init=True)
+class _Layout(Layout):
+    shape: NestedIntTuple
+    stride: NestedIntTuple
+
+    def __post_init__(self) -> None:
+        if not is_tuple(self.shape):
+            raise ValueError(f"shape must be a tuple, got {type(self.shape)}")
+        if not is_tuple(self.stride):
+            raise ValueError(f"stride must be a tuple, got {type(self.stride)}")
+        if len(flatten(self.shape)) != len(flatten(self.stride)):
+            raise ValueError(
+                f"sizes {len(flatten(self.shape))} and "
+                f"strides {len(flatten(self.stride))} must have the same length"
+            )
+
+    @property
+    def sizes(self) -> NestedIntTuple:
+        return self.shape
+
+    @property
+    def strides(self) -> NestedIntTuple:
+        return self.stride
+
+    @property
+    def sizes_and_strides(self) -> Iterator[tuple[int, int]]:
+        return zip(flatten(self.shape), flatten(self.stride))  # type: ignore[arg-type]
```

Comment:
OK, it's a lot clearer in this formulation of the PR. This method makes me nervous; it's not obvious to me that downstream people will use this correctly.